addons/account/models/account_move.py

```diff
@@ -3066,23 +3081,57 @@ def _get_last_sequence_domain(self, relaxed=False):
     def _get_starting_sequence(self):
         # EXTENDS account sequence.mixin
         self.ensure_one()
+        year_part = "%04d" % self.date.year
+        last_day = int(self.company_id.fiscalyear_last_day)
+        last_month = int(self.company_id.fiscalyear_last_month)
+        is_staggered_year = last_month != 12 or last_day != 31
+        if is_staggered_year:
+            if self.date > date(self.date.year, last_month, last_day):
+                year_part = "%s-%s" % (self.date.strftime('%y'), (self.date + relativedelta(years=1)).strftime('%y'))
+            else:
+                year_part = "%s-%s" % ((self.date + relativedelta(years=-1)).strftime('%y'), self.date.strftime('%y'))
         # Arbitrarily use annual sequence for sales documents, but monthly
         # sequence for other documents
         if self.journal_id.type in ['sale', 'bank', 'cash']:
-            starting_sequence = "%s/%04d/00000" % (self.journal_id.code, self.date.year)
+            # We reduce short code to 4 characters (0000) in case of staggered
+            # year to avoid too long sequences (see Indian GST rule 46(b) for
+            # example). Note that it's already the case for monthly sequences.
+            starting_sequence = "%s/%s/%s" % (self.journal_id.code, year_part, '0000' if is_staggered_year else '00000')
         else:
-            starting_sequence = "%s/%04d/%02d/0000" % (self.journal_id.code, self.date.year, self.date.month)
+            starting_sequence = "%s/%s/%02d/0000" % (self.journal_id.code, year_part, self.date.month)
         if self.journal_id.refund_sequence and self.move_type in ('out_refund', 'in_refund'):
             starting_sequence = "R" + starting_sequence
         if self.journal_id.payment_sequence and self.payment_id or self.env.context.get('is_payment'):
             starting_sequence = "P" + starting_sequence
         return starting_sequence
 
     def _get_sequence_date_range(self, reset):
+        if reset not in ('year_range', 'year_range_month'):
+            return super()._get_sequence_date_range(reset)
+
+        fiscalyear_last_day = self.company_id.fiscalyear_last_day
+        fiscalyear_last_month = self.company_id.fiscalyear_last_month
+        fiscal_year = date_utils.get_fiscal_year(self.date, day=fiscalyear_last_day, month=int(fiscalyear_last_month))
```

Comment:
well actually its the date_from, date_to of the fiscal year, it's a bit misleading

Reply:
I adapted to go with 2 explicit variables :+1: